code/modules/power/lgenerator.dm

```diff
@@ -11,355 +11,349 @@ TYPEINFO(/obj/machinery/power/lgenerator)
+					// Call proc to trigger rigged cell and log entries.
+
+	src.icon_state = "ggen[src.anchored]"
+
+	src.UpdateIcon()
+	return
+
+/obj/machinery/power/lgenerator/attack_hand(var/mob/user)
+	src.add_fingerprint(user)
+	src.ui_interact(user)
+
+/obj/machinery/power/lgenerator/ui_interact(mob/user, datum/tgui/ui)
+	ui = tgui_process.try_update_ui(user, src, ui)
+	if(!ui)
+		ui = new(user, src, "LocalGenerator")
+		ui.open()
+
+/obj/machinery/power/lgenerator/ui_data(mob/user)
+	. = list(
+		"name" = src.name,
+		"holding" = null, //represents the internal tank, the tank stats predefined UI block expects the name "holding"
+		"internalCell" = null,
+		"connectedAPC" = null,
+		"chargeAPC" = src.chargeAPC,
+		"boltsStatus" = src.anchored,
+		"generatorStatus" = src.active,
+	)
+	if(src.internalTank)
+		. += list(
+			"holding" = list(
+				"name" = src.internalTank.name,
+				"pressure" = MIXTURE_PRESSURE(src.internalTank.air_contents),
+				"maxPressure" = PORTABLE_ATMOS_MAX_RELEASE_PRESSURE,
+			)
+		)
+	if(src.internalCell)
+		. += list(
+			"internalCell" = list(
+				"name" = src.internalCell.name,
+				"chargePercent" = src.internalCell.percent(),
+			)
+		)
+	if(src.our_APC)
+		. += list(
+			"connectedAPC" = list(
+				"name" = src.our_APC.name,
+				"chargePercent" = src.our_APC.cell.percent(),
+			)
+		)
+
+/obj/machinery/power/lgenerator/ui_act(action, params, datum/tgui/ui)
+	. = ..()
+	if (.)
 		return
-
-	Topic(href, href_list)
-		if (!isturf(src.loc)) return
-		if (usr.getStatusDuration("stunned") || usr.getStatusDuration("weakened") || usr.stat || usr.restrained()) return
-		if (!issilicon(usr) && !in_interact_range(src, usr)) return
-
-		src.add_fingerprint(usr)
-		src.add_dialog(usr)
-
-		if (href_list["eject"])
-			if (src.active)
-				usr.show_text("Turn the generator off first!", "red")
-				return
-			if (src.P)
-				src.visible_message("<span class='alert'>[usr] ejects [src.P] from the [src]!</span>")
-				src.eject_tank(usr ? usr : null)
-			else
-				usr.show_text("There's no tank to eject.", "red")
-
-		if (href_list["eject-c"])
-			if (src.active && src.mode == 2)
-				usr.show_text("Turn the generator off first!", "red")
-				return
-			if (src.CL)
-				src.visible_message("<span class='alert'>[usr] ejects [src.CL] from the [src]!</span>")
-				src.eject_cell(usr ? usr : null)
-			else
-				usr.show_text("There's no cell to eject.", "red")
-
-		if (href_list["getAPC"])
-			switch (src.APC_check())
-				if (0)
-					src.our_APC = null
-					usr.show_text("Unable to establish connection to local APC.", "red")
-				if (1)
-					src.our_APC = get_local_apc(src)
-					usr.show_text("Connection to local APC established.", "blue")
-				if (2)
-					src.our_APC = null
-					usr.show_text("Local APC doesn't have a power cell to charge.", "red")
-				else
-					src.our_APC = null
-					usr.show_text("An error occurred, please try again.", "red")
-
-		if (href_list["togglebolts"])
+	. = TRUE
```

Comment:
Prefer setting this where something's been done, rather than doing it generally. May seem like unnecessary duplication, but maintains general coder understanding of what this is for rather than moving it away from where it's "really" used (i.e. typically at the end of switch cases that have done something that warrants a UI update).